service/frontend/workflowHandler.go

```diff
@@ -1734,10 +1734,12 @@ func (wh *WorkflowHandler) GetWorkflowExecutionHistory(
 		getRequest.MaximumPageSize = common.Int32Ptr(common.GetHistoryMaxPageSize)
 	}
 
-	enableArchivalRead := wh.GetArchivalMetadata().GetHistoryConfig().ReadEnabled()
-	historyArchived := wh.historyArchived(ctx, getRequest, domainID)
-	if enableArchivalRead && historyArchived {
-		return wh.getArchivedHistory(ctx, getRequest, domainID, scope)
+	if !getRequest.GetSkipArchival() {
+		enableArchivalRead := wh.GetArchivalMetadata().GetHistoryConfig().ReadEnabled()
+		historyArchived := wh.historyArchived(ctx, getRequest, domainID)
```

Comment:
Not related to your change but I think if `historyArchived` is true and either `skipArchival` or `enabledArchivalRead` is false, we can return `EntityNotExistError` directly. There's no need to `PollMutableState` from history host again.